cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/BrokerSetAwareGoal.java

```diff
@@ -0,0 +1,331 @@
+   */
+  @Override
+  protected void updateGoalState(ClusterModel clusterModel, OptimizationOptions optimizationOptions) throws OptimizationFailureException {
+    // Sanity check: No self-healing eligible replica should remain at a dead broker/disk.
+    GoalUtils.ensureNoOfflineReplicas(clusterModel, name());
+    // Sanity check: No replica should be moved to a broker, which used to host any replica of the same partition on its broken disk.
+    GoalUtils.ensureReplicasMoveOffBrokersWithBadDisks(clusterModel, name());
+    // Sanity check to confirm that the final distribution is broker set aware.
+    ensureBrokerSetAware(clusterModel, optimizationOptions);
+    if (_provisionResponse.status() != ProvisionStatus.OVER_PROVISIONED) {
+      _provisionResponse = new ProvisionResponse(ProvisionStatus.RIGHT_SIZED);
+    }
+    finish();
+  }
+
+  private void ensureBrokerSetAware(ClusterModel clusterModel, OptimizationOptions optimizationOptions)
+      throws OptimizationFailureException {
+    // Sanity check to confirm that the final distribution is brokerSet aware.
+    for (Map.Entry<String, List<Partition>> partitionsByTopic : clusterModel.getPartitionsByTopic().entrySet()) {
+      String topicName = partitionsByTopic.getKey();
+      if (!_excludedTopics.contains(topicName)) {
+        List<Partition> partitions = partitionsByTopic.getValue();
+        Set<Broker> allBrokersForTopic = partitions.stream()
+                                                   .map(partition -> partition.partitionBrokers())
+                                                   .flatMap(brokers -> brokers.stream())
+                                                   .collect(Collectors.toSet());
+        // Checks if a topic's brokers do not all live in a single brokerSet
+        if (_brokersByBrokerSet.values().stream().noneMatch(brokerSetBrokers -> brokerSetBrokers.containsAll(allBrokersForTopic))) {
+          throw new OptimizationFailureException(
+              String.format("[%s] Topic %s is not brokerSet-aware. brokers (%s).", name(), topicName, allBrokersForTopic));
+        }
+      }
+    }
+  }
+
+  /**
+   * brokerSet-awareness violations can be resolved with replica movements.
+   *
+   * @param broker              Broker to be balanced.
+   * @param clusterModel        The state of the cluster.
+   * @param optimizedGoals      Optimized goals.
+   * @param optimizationOptions Options to take into account during optimization.
+   */
+  @Override
+  protected void rebalanceForBroker(Broker broker, ClusterModel clusterModel, Set<Goal> optimizedGoals,
+                                    OptimizationOptions optimizationOptions) throws OptimizationFailureException {
+    LOG.debug("balancing broker {}, optimized goals = {}", broker, optimizedGoals);
+
+    String currentBrokerSetId = _brokerSetResolutionHelper.brokerSetId(broker);
+
+    for (Replica replica : broker.trackedSortedReplicas(replicaSortName(this, false, false)).sortedReplicas(true)) {
```

Comment:
In the latest testing, I found the _mustHaveTopicLeadersPerBroker actually participate in the replica movement, which is not expected. I'm not sure why it was not caught by my previous testing, but I'll take some time on why that happens and how to fix it.

Reply:
thanks @CCisGG. That is interesting. Let me know once you have idea on how to fix it.

Reply:
Hi @mohitpali , issue found. 

It's expected that the _mustHaveTopicLeadersPerBroker replicas to move if those replicas are originally from offline brokers. (See selectReplicasBasedOnExcludedTopics method.) That's why we didn't catch it before, as there were no offline brokers in previous rebalances.

The problem here is, if _mustHaveTopicLeadersPerBroker replicas moves, it can violate the MinTopicLeadershipPerBroker goal and throw optimization failure. To avoid that, MinTopicLeadershipPerBroker goal should be able to move those replicas, and the movement should be accepted by this BrokerSetAwareGoal.

The missing part of this PR is, we should honor _mustHaveTopicLeadersPerBroker or all _excludedTopics in actionAcceptance method of the BrokerSetAwareGoal. E.g:

```
    if (_excludedTopics.contains(action.topic())) {
      return ACCEPT;
    }
```

Reply:
1. With above logic we are saying that if any other goal moves excluded topic's replica, BrokerSetAwareGoal would accept the action. The excluded topic's replica won't be attempted to bring inside BrokerSet boundary.

Shall we limit it to _mustHaveTopicLeadersPerBroker only. That way other excluded topics's replicas, if offline, would still move within the BrokerSet.

Let me know what do you think. 

2. Have you tested your use case after making the above change locally, or should i make the change so that you can validate ?

Reply:
I have made the change and tested it locally. It works fine so far (no exceptions during rebalance).

For 1: That's a good question. After rethink about it, I believe make it _mustHaveTopicLeadersPerBroker actually makes more sense. Let's change it to _mustHaveTopicLeadersPerBroker

Reply:
Updated